### ISSUE TITLE
GCS upload timeout / spinning in poll

### ISSUE THREAD

**Issue by andir** [googleapis/google-cloud-cpp#3029]
### Does this issue affect the `google-cloud-cpp` project?

Yes, we are using the `google-cloud-cpp` library to upload files.

### What component of `google-cloud-cpp` is this related to?

- Cloud Storage specifically uploading new files.

### What version of `google-cloud-cpp` are you using?

0.12.0 & 0.11.0

### What compiler and version are you using?

gcc 7.4.0

### What operating system and version are you using?

NixOS with based on commit https://github.com/nixos/nixpkgs/commits/3f4144c30a6

(curl version 7.65.3, manually bumped google-cloud-cpp to 0.12.0 from 0.11.0 but same issue persists)

### What were you trying to do?

Upload files to a GCS bucket using [`nix`](https://nixos.org/nix). Specifically with the changes from my PR https://github.com/NixOS/nix/pull/3021.

We tried with big and small files. Both `InsertObject` and `WriteObject` and in all cases we get the same result (randomly). The relevant code is are at https://github.com/NixOS/nix/pull/3021/files#diff-b1e603a1ea8f9c5a0d38f7caf7079cdaR97-R122.

Command that we used (after building nix from that PR with a recent nixpkgs checkout):
```
$ nix copy --to gs://my-bucket-name?secret-key=/my/secret/signing/key /nix/store/…-something
```

### What did you expect to see?

To successfully upload the binary without wasting a lot of time in `poll(2)` and without seeing the error message in the end.

### What did you see instead?

`nix` did waste a lot of CPU time (minutes) in `poll(2)` called by `Curl_poll(2)`. See backtrace and perf data below.

Eventually the google-cloud-cpp library did throw an exception showing a `Request Timeout` error.

It seems like it never tells google that the upload is finished but waits for a response to the request.

### Anything else you would like us to know?

I am currently working on integrating GCS support into `nix`. While I can not see any obvious errors with my implementation this error occurs randomly for some people.

We captured a backtrace while the process was spinning in `poll(2)`:

```
Thread 1 (Thread 0x7f0418b7fa00 (LWP 15471)):
#0  0x00007f0419ba8501 in poll () from target:/nix/store/iykxb0bmfjmi7s53kfg6pjbfpd8jmza6-glibc-2.27/lib/libc.so.6
#1  0x00007f0419639399 in Curl_poll () from target:/nix/store/n8h43z4dh14hzqlvs7166sg53s1l6m2l-curl-7.65.3/lib/libcurl.so.4
#2  0x00007f0419632a8f in Curl_multi_wait.part.9 () from target:/nix/store/n8h43z4dh14hzqlvs7166sg53s1l6m2l-curl-7.65.3/lib/libcurl.so.4
#3  0x00007f041962c723 in curl_easy_perform () from target:/nix/store/n8h43z4dh14hzqlvs7166sg53s1l6m2l-curl-7.65.3/lib/libcurl.so.4
#4  0x00007f041906d7d4 in google::cloud::storage::v1::internal::CurlHandle::EasyPerform (this=0x7fff89bc3028) at /build/source/google/cloud/storage/internal/curl_handle.h:161
#5  google::cloud::storage::v1::internal::CurlRequest::MakeRequest (this=this@entry=0x7fff89bc2f70, payload=...) at /build/source/google/cloud/storage/internal/curl_request.cc:30
#6  0x00007f0419073538 in google::cloud::storage::v1::internal::CurlClient::UploadChunk (this=0xb41580, request=...) at /build/source/google/cloud/storage/internal/curl_client.cc:306
#7  0x00007f04190a0b7c in google::cloud::storage::v1::internal::CurlResumableUploadSession::UploadFinalChunk (this=0xd37aa0, buffer=..., upload_size=45744) at /build/source/google/cloud/storage/internal/curl_resumable_upload_session.cc:34
#8  0x00007f04190c578c in google::cloud::storage::v1::internal::LoggingResumableUploadSession::UploadFinalChunk (this=<optimized out>, buffer=..., upload_size=45744) at /build/source/google/cloud/storage/internal/logging_resumable_upload_session.cc:42
#9  0x00007f0419121dcd in google::cloud::storage::v1::internal::RetryResumableUploadSession::UploadFinalChunk (this=0xd3b880, buffer=..., upload_size=45744) at /build/source/google/cloud/storage/internal/retry_resumable_upload_session.cc:68
#10 0x00007f04190f0d0b in google::cloud::storage::v1::internal::ObjectWriteStreambuf::FlushFinal (this=this@entry=0xc6aeb0) at /build/source/google/cloud/storage/internal/object_streambuf.cc:297
#11 0x00007f04190f29a9 in google::cloud::storage::v1::internal::ObjectWriteStreambuf::Close (this=<optimized out>) at /build/source/google/cloud/storage/internal/object_streambuf.cc:226
#12 0x00007f041914d44b in google::cloud::storage::v1::ObjectWriteStream::CloseBuf (this=this@entry=0x7fff89bc4b00) at /build/source/google/cloud/storage/object_stream.cc:87
#13 0x00007f0419150f90 in google::cloud::storage::v1::ObjectWriteStream::Close (this=0x7fff89bc4b00) at /build/source/google/cloud/storage/object_stream.cc:83
#14 0x00007f041a3de33f in nix::GCSBinaryCacheStore::uploadFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#15 0x00007f041a3df8f7 in nix::GCSBinaryCacheStore::upsertFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#16 0x00007f041a33a860 in nix::BinaryCacheStore::addToStore(nix::ValidPathInfo const&, nix::ref<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, nix::RepairFlag, nix::CheckSigsFlag, std::shared_ptr<nix::FSAccessor>) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#17 0x00007f041a4569e8 in nix::Store::addToStore(nix::ValidPathInfo const&, nix::Source&, nix::RepairFlag, nix::CheckSigsFlag, std::shared_ptr<nix::FSAccessor>) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#18 0x00007f041a45f60d in nix::copyStorePath(nix::ref<nix::Store>, nix::ref<nix::Store>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, nix::RepairFlag, nix::CheckSigsFlag) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#19 0x00007f041a45f988 in nix::copyPaths(nix::ref<nix::Store>, nix::ref<nix::Store>, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, nix::RepairFlag, nix::CheckSigsFlag, nix::SubstituteFlag)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#3}::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#20 0x00007f041a45fc81 in std::_Function_handler<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), nix::copyPaths(nix::ref<nix::Store>, nix::ref<nix::Store>, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, nix::RepairFlag, nix::CheckSigsFlag, nix::SubstituteFlag)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#3}>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#21 0x00000000004fb1ba in std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const ()
#22 0x00007f041a461ced in nix::processGraph<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(nix::ThreadPool&, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::function<std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#23 0x00007f041a4621f0 in std::_Function_handler<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), nix::processGraph<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(nix::ThreadPool&, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::function<std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#24 0x00000000004fb1ba in std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const ()
#25 0x00007f041a460668 in std::_Function_handler<void (), std::_Bind<std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> (std::reference_wrapper<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const>)> >::_M_invoke(std::_Any_data const&) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#26 0x0000000000479cee in std::function<void ()>::operator()() const ()
#27 0x00007f041a24ec76 in nix::ThreadPool::doWork(bool) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixutil.so
#28 0x00007f041a24f87a in nix::ThreadPool::process() () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixutil.so
#29 0x00007f041a461318 in void nix::processGraph<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(nix::ThreadPool&, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::function<std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#30 0x00007f041a45db20 in nix::copyPaths(nix::ref<nix::Store>, nix::ref<nix::Store>, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, nix::RepairFlag, nix::CheckSigsFlag, nix::SubstituteFlag) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixstore.so
#31 0x00000000004b7912 in CmdCopy::run(nix::ref<nix::Store>, std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) ()
#32 0x00000000004b4f5d in nix::StorePathsCommand::run(nix::ref<nix::Store>) ()
#33 0x00000000004b2881 in nix::StoreCommand::run() ()
#34 0x00000000004ce597 in nix::mainWrapped(int, char**) ()
#35 0x00000000004ce7d8 in std::_Function_handler<void (), main::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#36 0x0000000000479cee in std::function<void ()>::operator()() const ()
#37 0x00007f041a503194 in nix::handleExceptions(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<void ()>) () from target:/nix/store/yi7w4516mlkd21h6sap40qzn2xha7xns-nix-2.3pre6779_324a5dc9/lib/libnixmain.so
#38 0x00000000004cde85 in main ()
```

The error messages that was eventually thrown was:

```
post-build-hook: Uploading paths /nix/store/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-xxx-0.8.0.0-doc /nix/store/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-xxx-0.8.0.0
post-build-hook: error: GCS error uploading 'nar/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.nar.xz': Permanent error in InsertObjectMedia: <!DOCTYPE html>
post-build-hook: <html lang=en>
post-build-hook:   <meta charset=utf-8>
post-build-hook:   <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
post-build-hook:   <title>Error 408 (Request Timeout)!!1</title>
post-build-hook:   <style>
post-build-hook:     *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
post-build-hook:   </style>
post-build-hook:   <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
post-build-hook:   <p><b>408.</b> <ins>That’s an error.</ins>
post-build-hook:   <p>Your client has taken too long to issue its request.  <ins>That’s all we know.</ins>
post-build-hook:  [INVALID_ARGUMENT]
error: program '/nix/store/mkgkz2pixh1wljvvcvqn8hp6h36i3nc1-upload-to-cache' failed with exit code 1
```
Here is the corresponding `perf -n --stdio` output:

https://storage.googleapis.com/nix-gcs-debugging/gi-soup/report_stdio.txt

This is currently a blocker for https://github.com/NixOS/nix/pull/3021.

**Comment by coryan** [comment]
Thank you for the detailed bug report.  It looks as-if the server is failing to respond, more than as-if the library is failing to send the final chunk.  But the library should handle this case, so we most likely have a bug regardless.

Can you answer some further questions to help us reproduce this problem?

Are you doing a lot of uploads through the same client before this failure?  If so, how many, how much time between them?

Are you waiting a long time in between uploads?  Sometimes the server closes the connection, but we should reconnect in that case, just trying to see if that is the problem.

I noticed you set the `contentType`, can you tell us what it is set to?  The server side handles gzip files differently, that would also help with any repro.  BTW, it should be possible to set  the `ContentEncoding` and `ContentType` options for an upload.

Can you tell us how often this happens?  You said "randomly", but does it happen 1% of the time, 10% of the time, 50% of the time?  Just trying to make sure our repro actually looks like yours.

There is another outstanding bug that might be related (#2980).

**Comment by andir** [comment]
> Thank you for the detailed bug report. It looks as-if the server is failing to respond, more than as-if the library is failing to send the final chunk. But the library should handle this case, so we most likely have a bug regardless.

Thank you for the quick reply. I agree that it seems like a server error and the client should probably handle that without hicups.

> Can you answer some further questions to help us reproduce this problem?

Of course :-)

> 
> Are you doing a lot of uploads through the same client before this failure? If so, how many, how much time between them?

In the case where we observe the error only a few uploads were made.

We usually upload on build results that consists of a small `.narinfo` file and one to three `.nar.xz` files. The outputs vary in size. Some of them are 500k and then there is usually a bigger one with up to 100MB.

> Are you waiting a long time in between uploads? Sometimes the server closes the connection, but we should reconnect in that case, just trying to see if that is the problem.

We are applying LZMA compression to all the files. The compression is applied before calling `InsertObject` or `WriteObject`. Some of the compression tasks last about a minute or two. The client is not destructed between those uploads. If it has an internal cache/pool of connections than that would be reused.

> I noticed you set the `contentType`, can you tell us what it is set to? The server side handles gzip files differently, that would also help with any repro. BTW, it should be possible to set the `ContentEncoding` and `ContentType` options for an upload.

We are setting ` application/x-nix-nar` for all the files in `nar/` and ` text/x-nix-narinfo` for all the files in the root ending in `.narinfo`. There isn't much an encoding besides binary I guess. The `.narinfo` files are just plaintext (possibly restricted to ASCII).

You mentioned `contentType` and `ContentType`. What is the difference?
 
> Can you tell us how often this happens? You said "randomly", but does it happen 1% of the time, 10% of the time, 50% of the time? Just trying to make sure our repro actually looks like yours.

During our ~12h of debugging yesterday we probably hat a hit chance of about 20%. 

Let me know if you need anything else.

**Comment by coryan** [comment]
Thank you for the additional information.  There must be something else going on because we have tests that upload files in the 100MiB range running more or less continuously. 

> We are applying LZMA compression to all the files. The compression is applied before calling `InsertObject` or `WriteObject`. Some of the compression tasks last about a minute or two. The client is not destructed between those uploads. If it has an internal cache/pool of connections than that would be reused.

I think that may be it.  I do not believe our tests pause between uploads, and the server can close connections that become idle.

At least that is something we can try to reproduce.

> You mentioned `contentType` and `ContentType`. What is the difference?

Other than me not spelling things consistently?  `google::cloud::storage::ContentType` is a type that you can provide to `WriteObject()` this sets the `Content-Type` header.  The service then puts that in the `contentType` metadata field.  There is a similar type (`storage::ContentEncoding`) to set the `contentEncoding` URL parameter.

You can also set those directly in the `ObjectMetadata` (or `WithObjectMetadata`) C++ arguments.

Why two ways to do the same thing?  Because setting the full object metadata implies a multi-part upload, and one may want to save a few bytes and use just headers and URL parameters.

**Comment by flokli** [comment]
I checked the logs again - indeed the timeout occured in the second upload round, after some compressing action - so the server might have decided to close the connection in the meantime.

I assume with server you mean a backend server? I assume the `Error 408 (Request Timeout)` comes from a reverse proxy in front of it.

**Comment by coryan** [comment]
> I assume with server you mean a backend server?

Yes, I did.  Unfortunately I cannot share details about the architecture (I am sure we must have published something, but I cannot speak about it without seeing what we have published, sorry).

**Comment by flokli** [comment]
Sure, no worries - just wanted to confirm I understood that part right :-)

Please let me know once you have some patch to share - happy to give it a try!

**Comment by remyabel** [comment]
In the meantime, would you mind removing the compression from `upsertFile` to see what happens?

**Comment by flokli** [comment]
Hm, disabling compression to wait less in betwen didn't seem to do the trick.

I created an empty bucket as a new cache, and started building (and uploading) again.

It succeeded in uploading some derivations, but failed again after some time:

```
post-build-hook: Uploading paths /nix/store/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-bash
post-build-hook: error: GCS error uploading 'nar/yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy.nar': Permanent error in UploadFinalChunk: <!DOCTYPE html>
post-build-hook: <html lang=en>
post-build-hook:   <meta charset=utf-8>
post-build-hook:   <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
post-build-hook:   <title>Error 408 (Request Timeout)!!1</title>
post-build-hook:   <style>
post-build-hook:     *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
post-build-hook:   </style>
post-build-hook:   <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
post-build-hook:   <p><b>408.</b> <ins>That’s an error.</ins>
post-build-hook:   <p>Your client has taken too long to issue its request.  <ins>That’s all we know.</ins>
post-build-hook:  [INVALID_ARGUMENT]
error: program '/nix/store/878v8a3h17ml84vh42afgazva90w5xbc-upload-to-cache' failed with exit code 1
```


This was now a version only using the streaming path:

https://github.com/andir/nix/commit/16a19e16e18426ec5685d1788c3fc1d16f850f07

**Comment by flokli** [comment]
With compression enabled, I hit a segfault (with coredump) - not sure if related, but there were quite some threads busy with curl and poll:

Backtrace: https://pastebin.com/FrwwixkG

**Comment by coryan** [comment]
> With compression enabled, I hit a segfault (with coredump) - not sure if related, but there were quite some threads busy with curl and poll:

We do not create threads in the GCS client library, nor does libcurl AFAIK.  Is the application creating threads, maybe in `nix::ThreadPool`?

If so, approximately how many?  And do they share the `storage::Client`?  Keep in mind, I still think this is a library bug, but that information helps us design a proper test to reproduce it.

**Comment by andir** [comment]
> > With compression enabled, I hit a segfault (with coredump) - not sure if related, but there were quite some threads busy with curl and poll:
> 
> We do not create threads in the GCS client library, nor does libcurl AFAIK. Is the application creating threads, maybe in `nix::ThreadPool`?
> If so, approximately how many? And do they share the `storage::Client`? Keep in mind, I still think this is a library bug, but that information helps us design a proper test to reproduce it.

Yes, threads are being created. Currently up to `std::thread::hardware_concurrency()` many threads. (@flokli: can you give us the hardware spec of your test VM?)
As far as I can tell from the top of my head all of them share the same `storage::Client` but not all threads are involved in uploading files.

Uploads are done in a way that keeps referential integrity. So if we build something and that depends on 5 libraries those libraries will uploaded before the final build result is being uploaded. Each of those libraries follows the same pattern. If multiple outputs that are queued for uploading have no pending dependencies then those will/can be uploaded in parallel.

The code for the graph processing can be found here: https://github.com/NixOS/nix/blob/fda7b95cb08c447a7ee4ec18ea3574d76e6264df/src/libutil/thread-pool.hh#L68
The way it is being used in our tests can be found here: https://github.com/NixOS/nix/blob/fda7b95cb08c447a7ee4ec18ea3574d76e6264df/src/libstore/store-api.cc#L613

**Comment by flokli** [comment]
The instance is a `n1-standard-8`, so it has 8 logical cores, and I see 16 threads in the stacktrace (8 of them doing GCS-related things).

**Comment by coryan** [comment]
I have not had a chance to write a test to reproduce this, but #2993 (part of the 0.13.0 release) has a good chance of fixing this problem.  Is there any chance you can test it?

**Comment by flokli** [comment]
I'll try to give this a go till the end of this week.

**Comment by coryan** [comment]
For what is worth.  I just sent PR #3066, it has an attempt to reproduce your problem, unfortunately it does not work as a repro. I was able to run 25 copies of the program at the same time, each copy running 4 threads, each thread doing 100 uploads and it just completes fine.

Maybe you can take a look and tell me if there is something about your application that I did not capture in the repro?

Also note that the `StreamingByChunk` test consumes about 8 times less memory than the `StreamingAll` test.

**Comment by coryan** [comment]
Any further thoughts?  I would really like to fix this, but if we cannot reproduce the problem I just will need to close the issue.

**Comment by flokli** [comment]
Sorry, I was quite busy - I'll try to get to this this week.

**Comment by flokli** [comment]
Bumping `google-cloud-cpp` to a recent version was a bit more work than I hoped - finished that in https://github.com/NixOS/nixpkgs/pull/70485, and will test GCS upload with this PR and google-cloud-cpp 0.14.0 tomorrow :-)

**Comment by flokli** [comment]
On a quick glimpse, I wasn't able to trigger the issue anymore, at least not with the current setup. I can do some more extensive testing mid of next week, but feel free to close for now.

However, I ran into another issue, maybe something not yet fixed by https://github.com/googleapis/google-cloud-cpp/pull/3062 - I opened https://github.com/googleapis/google-cloud-cpp/issues/3184 for that.

**Comment by coryan** [comment]
Thanks for looking into it. Closing, will continue the discussion on #3184.

**Comment by andir** [comment]
Just as an update:

With the latest version of this library and a dedicated client (with no connection pooling) per upload I am still running into the 408s. I was able to upload the very same files (to a different bucket) with the S3 compatibility enabled on the GCS site. I am starting to think that at least the timeouts are definitely something funny going on the backend.

I haven't had any crashes we this setup. Also leading me to the conclusion that some kind of concurrency issue is happening in the library despite the use of the supposedly threadsafe cURL APIs.

**Comment by coryan** [comment]
> @coryan
> Just as an update:
> 
> With the latest version of this library and a dedicated client (with no connection pooling) per upload I am still running into the 408s.

FWIW, 408s indicate that the server timed out a request because the client failed to send the complete request quickly enough.  Not sure that that means in this context. Possibly it means the client is starving threads and they cannot make progress or possibly it means you are trying to send a very large upload in a single request (from what I recall of the code that is unlikely).

> I was able to upload the very same files (to a different bucket) with the S3 compatibility enabled on the GCS site. I am starting to think that at least the timeouts are definitely something funny going on the backend.

Maybe.  It would not be the first time.  It is also possible you are running into something like: 

https://cloud.google.com/storage/docs/request-rate#auto-scaling

> 
> I haven't had any crashes we this setup. Also leading me to the conclusion that some kind of concurrency issue is happening in the library despite the use of the supposedly threadsafe cURL APIs.

It is entirely possible that the library has some concurrency problem.  It would not be the first time. It is *also* possible that it is libcurl (would not be the firs time either).  I had no luck debugging the problems for your setup (but admittedly I am not that familiar with NixOS), any output from valgrind / ASAN would be greatly appreciated.

**Comment by mamiu** [comment]
I think this issue is unrelated to the `google-cloud-cpp` library and probably a bug on GCS. Because I'm getting the exact same error when uploading a file with `curl` over a slow internet connection:

```html
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 408 (Request Timeout)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>408.</b> <ins>That’s an error.</ins>
  <p>Your client has taken too long to issue its request.  <ins>That’s all we know.</ins>
```

But the strange thing is, that the file is being uploaded completely without any interruptions and at 100% it gets stuck and returns the error message above after a while.

**Comment by andir** [comment]
Yeah, after all this time I also believe it is GCS that is at fault. How
large are your files? I believe we are supposed to use the chunked
uploading and large single-stream uploads might simply not be used
enough and buggy.
